server/src/graql/gremlin/TraversalPlanner.java

```diff
@@ -110,33 +111,42 @@ public static GraqlTraversal createTraversal(Pattern pattern, TransactionOLTP tx
 
         // build a query plan for each query subgraph separately
         for (Set<Fragment> connectedFragments : connectedFragmentSets) {
-            // collect the mapping from directed bedge back to fragments -- inverse operation of creating virtual middle nodes
-            Map<Node, Map<Node, Fragment>> middleNodeFragmentMapping = virtualMiddleNodeToFragmentMapping(connectedFragments, queryGraphNodes);
-
+            // one of two cases - either we have a connected graph > 1 node, which is used to compute a MST, OR exactly 1 node
             Arborescence<Node> subgraphArborescence = computeArborescence(connectedFragments, queryGraphNodes, tx);
             if (subgraphArborescence != null) {
-                List<Fragment> subplan = ArborescenceToPlan.greedyTraversal(subgraphArborescence, queryGraphNodes, middleNodeFragmentMapping);
+                // collect the mapping from directed edge back to fragments -- inverse operation of creating virtual middle nodes
+                Map<Node, Map<Node, Fragment>> middleNodeFragmentMapping = virtualMiddleNodeToFragmentMapping(connectedFragments, queryGraphNodes);
+                List<Fragment> subplan = GreedyTreeTraversal.greedyTraversal(subgraphArborescence, queryGraphNodes, middleNodeFragmentMapping);
                 plan.addAll(subplan);
+            } else {
+                // find and include all the nodes not touched in the MST in the plan
+               Set<Node> unhandledNodes = connectedFragments.stream()
+                        .flatMap(fragment -> fragment.getNodes().stream())
+                        .map(node -> queryGraphNodes.get(node.getNodeId()))
+                        .collect(Collectors.toSet());
+               if (unhandledNodes.size() != 1) {
+                   throw GraknServerException.create("Query planner exception - expected one unhandled node, found " + unhandledNodes.size());
+               }
+               plan.addAll(nodeToPlanFragments(Iterators.getOnlyElement(unhandledNodes.iterator()), queryGraphNodes, false));
             }
+        }
 
-            // find and include all the nodes not touched in the MST in the plan
-            Set<Node> connectedNodes = connectedFragments.stream()
-                    .flatMap(fragment -> fragment.getNodes().stream())
-                    .map(node -> queryGraphNodes.get(node.getNodeId()))
-                    .collect(Collectors.toSet());
-            plan.addAll(fragmentsForUnvisitedNodes(queryGraphNodes, connectedNodes));
+        // this shouldn't be necessary, but we keep it just in case of an edge case that we haven't thought of
+        List<Fragment> remainingFragments = fragmentsForUnvisitedNodes(queryGraphNodes, queryGraphNodes.values());
+        if (remainingFragments.size() > 0) {
+            LOG.error("Expected all fragments to be handled, but found these: " + remainingFragments);
```

Comment:
Given that we don't interrupt the execution of the code if `remainingFragments.size() > 0` we might want to print the message as a WARN rather than ERROR.

I am assuming that this case does not undermine the correct execution of the QueryPlanner

Reply:
Ahh warn is what i was looking for - better than debug or error!